### PR TITLE
Tiers can be sorted by category

### PR DESCRIPTION
**Pull request by mejango**
This is turning into a meaningful PR:

Motivation:
There exist product opportunities to leverage different categories of NFTs that route proceeds into treasuries. Project owners want to be able to add new NFT tiers in different contexts to mean different things. In order to do so, we need to organize the tier data structures better to give clients and composing contracts access to NFTs on a per-category basis.

Category indexes, like JBProject metadata `domains`, can have their meaning standardized and understood off-protocol. i.e. JBM can set a standard that tiers of category 1 represent top-level payment perks, whereas category 2 represents NFT tiers posted in response to community activity. The contracts do not need to keep track of what each category represents or how they're ordered, since products can make them out to mean various things.

Code spec changes:
- Tiers are sorted by category first, then order of having been added. There's no more ordering based on contributionFloor. This is OK since we removed the mintBestAvailable functionality in v1.1.

- the `tiers` getter can now specify a category to get from. Send 0 to get from all categories. 

- internal variable names have been improved to better convey the structure.

- apart from a few slight API changes, there should be no functional changes assuming all tiers added belong to the same category.

Open questions:
- should we rename "category" to "domain" to better match the way JBProject metadata is stores, and JBSplits are stored? They functionally work the same way as just a means of namespace organization. cc @drgorillamd @viraj124 

Test needs:
- remove contributionFloor sorting based tests.
- adjust `tiers` API tests to be sent 0 as `category` to maintain current behavior, and test getting tiers from a specific category. cc @viraj124

### CODE REVIEW

**Review by viraj124 — APPROVED** [review]

**Comment by viraj124** [comment]
> This is turning into a meaningful PR:
> 
> Motivation: There exist product opportunities to leverage different categories of NFTs that route proceeds into treasuries. Project owners want to be able to add new NFT tiers in different contexts to mean different things. In order to do so, we need to organize the tier data structures better to give clients and composing contracts access to NFTs on a per-category basis.
> 
> Category indexes, like JBProject metadata `domains`, can have their meaning standardized and understood off-protocol. i.e. JBM can set a standard that tiers of category 1 represent top-level payment perks, whereas category 2 represents NFT tiers posted in response to community activity. The contracts do not need to keep track of what each category represents or how they're ordered, since products can make them out to mean various things.
> 
> Code spec changes:
> 
> * Tiers are sorted by category first, then order of having been ordered. There's no more ordering based on contributionFloor. This is OK since we removed the mintBestAvailable functionality in v1.1.
> * the `tiers` getter can now specify a category to get from. Send 0 to get from all categories.
> * variable names have been improved to better convey the structure.
> * apart from a few slight API changes, there should be no functional changes assuming all tiers added belong to the same category.
> 
> Open questions:
> 
> * should we rename "category" to "domain" to better match the way JBProject metadata is stores, and JBSplits are stored? They functionally work the same way as just a means of namespace organization. cc @drgorillamd @viraj124
> 
> Test needs:
> 
> * remove contributionFloor sorting based tests.
> * adjust `tiers` API tests to be sent 0 as `category` to maintain current behavior, and test getting category specific tiers. cc @viraj124

sounds good, in terms of naming my vote be with "category" in terms of the NFT context probably better for readability but no issues with "domain" too

File: contracts/forge-test/NFTReward_Unit.t.sol [hunk 1]
```diff
@@ -1970,6 +2007,137 @@ contract TestJBTieredNFTRewardDelegate is Test {
...
+    // Check: Expected number of tiers?
+    assertEq(_storedTiers.length, _tiers.length + _tiersAdded.length);
+
+    // Check: Are all tiers in the new tiers (unsorted)?
+    assertTrue(_isIn(_tiers, _storedTiers)); // Original tiers
+    assertTrue(_isIn(_tiersAdded, _storedTiers)); // New tiers
+
+    // Check: Are all the tiers sorted?
+    for (uint256 i = 1; i < _storedTiers.length; i++) {
+      assertLe(_storedTiers[i - 1].category, _storedTiers[i].category);
```
Comment (drgorillamd) [inline contracts/forge-test/NFTReward_Unit.t.sol:2137]: assertLt instead?

File: contracts/forge-test/NFTReward_Unit.t.sol [hunk 2]
```diff
@@ -264,8 +267,8 @@ contract TestJBTieredNFTRewardDelegate is Test {
 
     _delegate.transferOwnership(owner);
 
-    assertTrue(_isIn(_delegate.test_store().tiers(address(_delegate), 0, numberOfTiers), _tiers));
-    assertTrue(_isIn(_tiers, _delegate.test_store().tiers(address(_delegate), 0, numberOfTiers)));
+    assertTrue(_isIn(_delegate.test_store().tiers(address(_delegate), 0, 0, numberOfTiers), _tiers));
```
Comment (drgorillamd) [inline contracts/forge-test/NFTReward_Unit.t.sol:270]: can't we test the catergory too (instead of 0)?

**Review by drgorillamd — APPROVED** [review]

File: contracts/forge-test/NFTReward_Unit.t.sol (hunk 2 above)
Comment (viraj124) [reply contracts/forge-test/NFTReward_Unit.t.sol:270]: We have a separate test for that to fetch a particular category check the last commit

**Pull request merged**